// @lynx-js/react@0.114.0/runtime/lib/pendingListUpdates.js
// Copyright 2025 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
export const __pendingListUpdates = {
    values: {},
    clear(id) {
        if (this.values) {
            delete this.values[id];
        }
    },
    clearAttachedLists() {
        if (this.values) {
            Object.values(this.values)
                .map(update => update.getAttachedListId())
                .filter(id => id !== undefined)
                .forEach(id => this.clear(id));
        }
    },
    flush() {
        if (this.values) {
            Object.values(this.values)
                .map(update => update.flush())
                .filter(id => id !== undefined)
                .forEach(id => this.clear(id));
        }
    },
    flushWithId(id) {
        if (this.values && this.values[id]?.flush() !== undefined) {
            this.clear(id);
        }
    },
    runWithoutUpdates(cb) {
        const old = this.values;
        this.values = null;
        try {
            cb();
        }
        finally {
            this.values = old;
        }
    },
};
